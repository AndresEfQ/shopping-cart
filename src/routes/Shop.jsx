import { useEffect, useState } from "react";
import styled from "styled-components";
import SimpleBar from "simplebar-react";
import "simplebar-react/dist/simplebar.min.css"
import { FaSearch } from "react-icons/fa";
import DotLoader from "react-spinners/DotLoader";
import CardsList from "../components/CardsList";
import backupData from "../assets/cache";

export default function Shop({windowWidth}) {

  const [apiResponse, setApiResponse] = useState();
  const [sets, setSets] = useState();
  const [cards, setCards] = useState(backupData.cards);
  const [isLoadingCards, setIsLoadingCards] = useState(false);
  const [searchInput, setSearchInput] = useState("");

  useEffect(() => {

    fetch("https://api.magicthegathering.io/v1/sets")
    .then(response => response.json())
    .then(response => {
      setApiResponse(response.sets);
      setSets(response.sets);
    })
    .catch(err => console.log(err));

  }, []);

  const searchSets = (query, array) => {
    return array.filter(element => element.name.toLowerCase().indexOf(query.toLowerCase()) !== -1);
  }
  
  const handleSearchInputChange = (e) => {
    setSearchInput(e.target.value);
    setSets(searchSets(e.target.value, apiResponse));
  }

  const handleSelectSet = (e) => {
    e.preventDefault();
    setIsLoadingCards(true);
    const set = e.target.id || e.target[0].value
    console.log(e.target[0].value)
    fetch(`https://api.magicthegathering.io/v1/cards?set=${set}&random=true&pageSize=12`)
    .then(response => response.json())
    .then(response => {
      setCards(response.cards);
      setIsLoadingCards(false);
      console.log(response.cards)
    })
    .catch(err => console.log(err));
  }

  return (
    <ShopDiv>
      {(windowWidth > 490) ? 
        <div>
          <SearchBar
            type="search"
            placeholder="Search Sets"
            onChange={handleSearchInputChange}
            value={searchInput}
          />
          <FaSearch color="var(--grey)"/>
          <SideBar>
            <SimpleBar style={{maxHeight: "80vh", width: "15vw"}}>
              {sets
                ? sets.map(set => {
                  return (
                    <SetSelector
                      key={set.code}
                      id={set.code}
                      onClick={handleSelectSet}>{set.name}
                    </SetSelector>
                  )})
                : <span>loading</span>}
            </SimpleBar>
          </SideBar>
        </div> : 
        <form onSubmit={handleSelectSet}>
          <label htmlFor="sets">Choose a set:</label>
          <input list="setsList" id="sets" name="sets" />
          <datalist id="setsList">
            {sets && sets.map(set => {
              return (
                <option
                  key={set.code}
                  data-code={set.code}
                  value={set.code}
                >{set.name}</option>
              )})}
          </datalist>
          <button>OK</button>
        </form>
      }
      {cards ? <CardsList cards={cards} /> : <Shade><DotLoader color="var(--main)" /></Shade>}
      {isLoadingCards && <Shade><DotLoader color="var(--main)" /></Shade>}            
    </ShopDiv>
  )
}

const ShopDiv = styled.div`

  position: relative;
  display: flex;

  @media only screen and (max-width: 490px) {
    flex-direction: column;
    width: 100%
  }

  & > div > svg {
    position: absolute;
    top: 2.6vh;
    left: 2rem;
  }

  & > input {
    width: 60%;
  }
`

const SideBar = styled.aside`
  height: 80vh;
  width: 15vw;
  margin-left: 1rem;
  margin-top: 2vh;
  position: relative;
  display: flex;
  flex-direction: column;
  overflow: auto;
  background-color: var(--secondary-op60);
  scrollbar-width: none;
  -ms-overflow-style: none;
`;

const SearchBar = styled.input`
  margin-left: 1.5rem;
  margin-top: 2vh;
  padding: 0.2rem 1.8rem;
  width: 13vw;
`;

const SetSelector = styled.button`
  width: 14vw;
  margin: 0.5vh;
  padding: 0.4rem 0.8rem;
  overflow-x: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: left;
  cursor: pointer;
`;

const Shade = styled.div`
  height: 88vh;
  width: 100vw;
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  background: var(--op80);
  z-index: 3;
`;
