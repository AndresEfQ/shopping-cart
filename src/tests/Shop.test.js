import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Shop from "../routes/Shop";

beforeEach(() => {
  fetch.resetMocks();
  fetch.mockResponseOnce(JSON.stringify({sets: []}));
});

describe("Shop component", () => {

  test("Renders correctly", () => {
    const { container } = render(<Shop />);
    expect(container).toMatchSnapshot();
  })

  test("Fetches data on mount", async () => {
    render(<Shop />)
    expect(fetch).toHaveBeenCalledTimes(1);
  });
});