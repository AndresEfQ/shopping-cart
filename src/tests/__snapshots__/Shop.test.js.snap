// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop component Renders correctly 1`] = `
<div>
  <div
    class="sc-beqWaB dNCgcm"
  >
    <input
      class="sc-dmqHEX iVPWa-d"
      placeholder="Search Sets"
      type="search"
    />
    <svg
      color="#777"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      style="color: rgb(119, 119, 119);"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
      />
    </svg>
    <aside
      class="sc-gueYoa jYDwFW"
    >
      <div
        data-simplebar="init"
        style="max-height: 80vh; width: 15vw;"
      >
        <div
          class="simplebar-wrapper"
        >
          <div
            class="simplebar-height-auto-observer-wrapper"
          >
            <div
              class="simplebar-height-auto-observer"
            />
          </div>
          <div
            class="simplebar-mask"
          >
            <div
              class="simplebar-offset"
              style="right: 0px; bottom: 0px;"
            >
              <div
                aria-label="scrollable content"
                class="simplebar-content-wrapper"
                role="region"
                style="height: auto; overflow-x: hidden; overflow-y: hidden;"
                tabindex="0"
              >
                <div
                  class="simplebar-content"
                >
                  <span>
                    loading
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="simplebar-placeholder"
            style="width: 0px; height: 0px;"
          />
        </div>
        <div
          class="simplebar-track simplebar-horizontal"
          style="visibility: hidden;"
        >
          <div
            class="simplebar-scrollbar"
            style="width: 0px; display: none;"
          />
        </div>
        <div
          class="simplebar-track simplebar-vertical"
          style="visibility: hidden;"
        >
          <div
            class="simplebar-scrollbar"
            style="height: 0px; display: none;"
          />
        </div>
      </div>
    </aside>
  </div>
</div>
`;
